### ISSUE TITLE
Enable Google::Cloud.configure for all Google Cloud gems

### ISSUE THREAD

**Issue by blowmage** [GoogleCloudPlatform/google-cloud-ruby#1803]
The Stackdriver gems have introduced the following additional configuration syntax:

```ruby
Google::Cloud.configure do |config|
  config.debugger.project_id = "YOUR-PROJECT-ID"
  config.debugger.keyfile    = "/path/to/service-account.json"
end
```

This can be used to configure more than just authentication:

```ruby
Google::Cloud::ErrorReporting.configure do |config|
  config.project_id = "my-project-id"
  config.keyfile = "/path/to/keyfile.json"

  config.service_name = "my-service"
  config.service_version = "v8"
end
```

Should this be enabled for all Google Cloud gems?

**Comment by dazuma** [comment]
TL;DR: Maybe, if we think it's useful, but I'm not yet convinced it is. The need for it in the stackdriver gems is different (and much stronger) than any of the other client libraries.

Here's some background on the stackdriver gems configuration syntax that might help.

It is something @hxiong388 and I put together to provide a sane "global" configuration mechanism for the stackdriver gems. Those gems have an instrumentation component that relies on Rack middleware. Rails apps normally use the Rails configuration mechanism to configure middleware globally, but for Sinatra and other frameworks we used to require the user to provide all this information on middleware construction, which was extremely clumsy. Thus, we incorporated this configuration syntax into stackdriver-core for non-Rails apps. The important thing to note is that this is "global" configuration, used for (1) rack middleware, and (2) class methods such as `Google::Cloud::ErrorReporting.report` (that don't require you to create a project object explicitly, for convenience). Thus it requires a global configuration syntax like you see here.

We could use it to provide "defaults" (e.g. for the project and keyfile) that can be set once and shared across clients so that users don't have to pass them into every constructor. That might be useful, but it seems less important than the need we had for the stackdriver gems. Note also that we already use environment variables (e.g. `GOOGLE_CLOUD_KEYFILE`) for this purpose.

**Comment by hxiong388** [comment]
Agree with everything @dazuma mentioned above. But I think the biggest driver to make all the clients to also default from this configuration mechanism is consistency. I can imagine there are users current doing this in their Rails or Sinatra app:

```ruby
Google::Cloud.configure do |config|
  config.debugger.project_id = "YOUR-PROJECT-ID"
  config.debugger.keyfile    = "/path/to/service-account.json"
end

storage = Google::Cloud::Storage.new #=> project id missing error
```

New users will be very confused when they encounter this pitfall.

**Comment by blowmage** [comment]
The consistency argument is what is driving my view on it. I would prefer if Storage could take advantage of the config when set.

**Comment by blowmage** [comment]
To add to Heng's argument, the debugger README also has the following code example:

```ruby
require "google/cloud/debugger"
Google::Cloud.configure do |config|
  # Shared parameters
  config.project_id = "your-project-id"
  config.keyfile = "/path/to/key.json"
  # Or Stackdriver Debugger agent specific parameters
  config.debugger.project_id = "your-project-id"
  config.debugger.keyfile = "/path/to/key.json"
end
```

Those shared parameters are intended to be shared across all libraries, but currently are only shared for Stackdriver gems. I believe this is a problem and the config syntax should be supported on all Google Cloud gems.

**Comment by blowmage** [comment]
This is being worked on in #1912.

**Comment by blowmage** [comment]
Closed by #1912.
